<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''SELECT 
            COUNT(*) count_vacancies
            FROM vacancies
'''

In [5]:
# результат запроса
count_vacancies = pd.read_sql_query(query_3_1, connection)
count_vacancies

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\2627226083.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vacancies = pd.read_sql_query(query_3_1, connection)


count_vacancies
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = '''SELECT 
            COUNT(*) count_employers
            FROM employers
'''

In [7]:
# результат запроса
count_employers = pd.read_sql_query(query_3_2, connection)
count_employers

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\2124197464.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers = pd.read_sql_query(query_3_2, connection)


count_employers
0            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = '''SELECT 
            COUNT(*) count_areas
            FROM areas
'''

In [9]:
# результат запроса
count_areas = pd.read_sql_query(query_3_3, connection)
count_areas

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\4290886743.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_areas = pd.read_sql_query(query_3_3, connection)


count_areas
0         1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = '''SELECT 
            COUNT(*) count_industries
            FROM industries
'''

In [11]:
# результат запроса
count_industries = pd.read_sql_query(query_3_4, connection)
count_industries

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\2328395817.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_industries = pd.read_sql_query(query_3_4, connection)


count_industries
0               294

***

## Выводы по предварительному анализу данных
### В нашей базе содержится:
- Почти 50 тысяч вакансий;
- Более 20 тысяч работодателей;
- Вакансии предлагают в 1362 регионах;
- Почти 300 сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''SELECT
            a.name as area,
            COUNT(*) as cnt
            FROM areas as a
            JOIN vacancies v on a.id=v.area_id
            GROUP BY a.id
            ORDER BY cnt desc
            LIMIT 5
'''

In [13]:
# результат запроса
vacancies_count = pd.read_sql_query(query_4_1, connection)
vacancies_count

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\3851789608.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''SELECT
            COUNT(*) cnt
            FROM vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [15]:
# результат запроса
salary_count = pd.read_sql_query(query_4_2, connection)
salary_count

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\949394385.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_count = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''SELECT
            round(AVG(salary_from)) avg_salary_from,
            round(AVG(salary_to)) avg_salary_to
            FROM vacancies
'''

In [17]:
# результат запроса
avg_salary = pd.read_sql_query(query_4_3, connection)
avg_salary

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\212845365.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''SELECT
            COUNT(*) cnt,
            schedule,
            employment
            FROM vacancies
            GROUP BY 2,3
            ORDER BY 1 DESC
'''

In [19]:
# результат запроса
schedule_count_vacancies = pd.read_sql_query(query_4_4, connection)
schedule_count_vacancies

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\1267644835.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  schedule_count_vacancies = pd.read_sql_query(query_4_4, connection)


cnt          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''SELECT
            experience,
            COUNT(*)
            FROM vacancies
            GROUP BY 1
            ORDER BY 2 
'''

In [21]:
# результат запроса
experience_count = pd.read_sql_query(query_4_5, connection)
experience_count

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\1998695830.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience_count = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### Выводы по детальному анализу вакансий:

- Чем больше город, тем больше количество вакансий;
- У 24073 вакансий заполнено хотя бы одно из двух полей с зарплатой (нижняя или верхняя граница зарплатной вилки);
- Среднее значение для нижней границы зарплатной вилки - 71 тысяча рублей, верхней - 110 тысяч рублей;
- Самое большое количество вакансий доступно для людей с опытом работы от 1 года до 3 лет;
- Для специалистов, у которых опыт работы более 6 лет меньше всего вакансий;
- Самым популярным сочетанием "график работы - тип занятости" является "полный день - полная занятость".

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''SELECT
            e.name,
            COUNT(*)
            FROM employers e
            JOIN vacancies v ON v.employer_id = e.id
            GROUP BY 1
            ORDER BY 2 DESC
            LIMIT 5      
'''

In [23]:
# результат запроса
employers_name = pd.read_sql_query(query_5_1, connection)
employers_name

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\708251211.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_name = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''SELECT
               a.name,
               COUNT(e.id) cnt_emp,
               COUNT(v.id) cnt_vac
               FROM areas a
               LEFT JOIN employers e ON a.id = e.area
               LEFT JOIN vacancies v ON v.area_id = a.id
               WHERE v.id IS NULL
               GROUP BY 1
               ORDER BY 2  DESC 
'''

In [25]:
# результат запроса
emp_cnt = pd.read_sql_query(query_5_2, connection)
emp_cnt

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\3358796221.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_cnt = pd.read_sql_query(query_5_2, connection)


name  cnt_emp  cnt_vac
0                            Россия      410        0
1                         Казахстан      207        0
2                Московская область       75        0
3                Краснодарский край       19        0
4                          Беларусь       18        0
..                              ...      ...      ...
588                        Кувандык        0        0
589                        Куйбышев        0        0
590  Куйбышево (Ростовская область)        0        0
591                        Култаево        0        0
592                        Кульсары        0        0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''SELECT
                e.name,
                COUNT(DISTINCT(v.area_id)) cnt_area
                FROM employers e 
                LEFT JOIN vacancies v ON e.id = v.employer_id
                GROUP BY 1
                ORDER BY 2  DESC   
'''

In [27]:
# результат запроса
cnt_area = pd.read_sql_query(query_5_3, connection)
cnt_area

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\3588327737.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_area = pd.read_sql_query(query_5_3, connection)


name  cnt_area
0                      Яндекс       181
1                  Ростелеком       152
2                  Спецремонт       116
3      Поляков Денис Иванович        88
4                   ООО ЕФИН         71
...                       ...       ...
23170                  СДЕЛКА         0
23171            Alandr Group         0
23172                     СДК         0
23173      Lemon Land Lombard         0
23174                  Махина         0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''SELECT
                COUNT(e.id)
                FROM employers e 
                LEFT JOIN employers_industries ei ON e.id = ei.employer_id
                WHERE ei.industry_id IS NULL
'''

In [29]:
# результат запроса
emp_ind = pd.read_sql_query(query_5_4, connection)
emp_ind

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\2045131046.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_ind = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''SELECT
                e.name,
                COUNT(ei.industry_id) cnt
                FROM employers e 
                LEFT JOIN employers_industries ei ON e.id = ei.employer_id
                WHERE ei.industry_id IS NOT NULL
                GROUP BY 1
                HAVING COUNT(ei.industry_id) = 4
                ORDER BY 1
                OFFSET 2 LIMIT 1
'''

In [31]:
# результат запроса
employers_industry = pd.read_sql_query(query_5_5, connection)
employers_industry

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\1629656118.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_industry = pd.read_sql_query(query_5_5, connection)


name  cnt
0  2ГИС    4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''SELECT
                COUNT(e.id)
                FROM employers_industries ei
                JOIN employers e ON e.id = ei.employer_id
                JOIN industries i ON i.id = ei.industry_id
                WHERE i.name = 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
employers_industry_2 = pd.read_sql_query(query_5_6, connection)
employers_industry_2

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\1224284323.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_industry_2 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников

In [35]:
# текст запроса

In [36]:
# результат запроса

***

### Выводы по анализу работодателей:
- Лидерами по количеству вакансий являются: Яндекс, Ростелеком, Тинькофф, Сбер, Газпром.
- Наибольшее количество работодателей в России;
- Яндекс является лидером по количество вакансий в разных городах;
- У 8419 работодателей не указана сфера деятельности;
- В сфере "Разработки программного обеспечения" работает 3553 компании.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1 = f'''SELECT
                COUNT(v.id) data_vacancies
                FROM vacancies v
                WHERE lower(v.name) LIKE '%data%' OR lower(v.name) LIKE '%данн%'
'''

In [38]:
# результат запроса
ds_vacancies = pd.read_sql_query(query_6_1, connection)
ds_vacancies

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\1937916000.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_vacancies = pd.read_sql_query(query_6_1, connection)


data_vacancies
0            1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2 = f'''SELECT 
                COUNT(id) junior_vacancies
                FROM vacancies 
                WHERE (lower(name) LIKE '%data scientist%' 
                OR lower(name) LIKE '%data science%' 
                OR lower(name) LIKE '%исследователь данных%' 
                OR (name LIKE '%ML%' and name NOT LIKE '%HTML%') 
                OR lower(name) LIKE '%machine learning%' 
                OR lower(name) LIKE '%машинн%обучен%')
                AND 
                (lower(name) LIKE '%junior%' 
                OR lower(experience) LIKE '%нет опыта%' 
                OR lower(employment) LIKE '%стажировка%')
'''

In [40]:
# результат запроса
junior_vacancies = pd.read_sql_query(query_6_2, connection)
junior_vacancies

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\625073081.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  junior_vacancies = pd.read_sql_query(query_6_2, connection)


junior_vacancies
0                51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3 = f'''SELECT 
                COUNT(id) sql_skills
                FROM vacancies 
                WHERE (lower(name) LIKE '%data scientist%' 
                OR lower(name) LIKE '%data science%' 
                OR lower(name) LIKE '%исследователь данных%' 
                OR (name LIKE '%ML%' and name NOT LIKE '%HTML%') 
                OR lower(name) LIKE '%machine learning%' 
                OR lower(name) LIKE '%машинн%обучен%')
                AND (key_skills LIKE '%SQL%'
                OR lower(key_skills) LIKE  '%postgres%')              
'''

In [42]:
# результат запроса
sql_vacancies = pd.read_sql_query(query_6_3, connection)
sql_vacancies

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\3698489625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql_vacancies = pd.read_sql_query(query_6_3, connection)


sql_skills
0         201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f'''SELECT 
                COUNT(id) python_skills
                FROM vacancies 
                WHERE (lower(name) LIKE '%data scientist%' 
                OR lower(name) LIKE '%data science%' 
                OR lower(name) LIKE '%исследователь данных%' 
                OR (name LIKE '%ML%' and name NOT LIKE '%HTML%') 
                OR lower(name) LIKE '%machine learning%' 
                OR lower(name) LIKE '%машинн%обучен%')
                AND (lower(key_skills) LIKE  '%python%')
'''   

In [44]:
# результат запроса
python_skills = pd.read_sql_query(query_6_4, connection)
python_skills

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\950132531.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  python_skills = pd.read_sql_query(query_6_4, connection)


python_skills
0            351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5 = f'''SELECT 
                ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) avg_ds_skills
                FROM vacancies 
                WHERE (lower(name) LIKE '%data scientist%' 
                OR lower(name) LIKE '%data science%' 
                OR lower(name) LIKE '%исследователь данных%' 
                OR (name LIKE '%ML%' and name NOT LIKE '%HTML%') 
                OR lower(name) LIKE '%machine learning%' 
                OR lower(name) LIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
'''

In [46]:
# результат запроса
avg_ds_skills = pd.read_sql_query(query_6_5, connection)
avg_ds_skills

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\1109382575.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_ds_skills = pd.read_sql_query(query_6_5, connection)


avg_ds_skills
0           6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
query_6_6 = f'''SELECT 
                experience,
                ROUND(AVG(COALESCE((salary_to + salary_from)/2, salary_from, salary_to))) avg_salary
                FROM vacancies 
                WHERE (lower(name) LIKE '%data scientist%' 
                OR lower(name) LIKE '%data science%' 
                OR lower(name) LIKE '%исследователь данных%' 
                OR (name LIKE '%ML%' and name NOT LIKE '%HTML%') 
                OR lower(name) LIKE '%machine learning%' 
                OR lower(name) LIKE '%машинн%обучен%')
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                AND (experience IS NOT NULL)
                GROUP BY 1
'''

In [48]:
# результат запроса
avg_salary = pd.read_sql_query(query_6_6, connection)
avg_salary

C:\Users\Asus\AppData\Local\Temp\ipykernel_2632\4147299713.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

### Выводы по предметному анализу:
- 1771 вакансия среди общего количества вакансий имеет отношение к данным;
- Для начинающего дата-сайентиста есть 51 вакансия;
- Вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres - 201;
- Самые востребованные навыки это Python, SQL, Postgre;
- Средняя заработная плата специалистов без опыта составляет 75 тысяч рублей, с опытом работы от 3 до 6 лет - 240 тысяч рублей.

### Общий вывод по проекту
1. По результатам исследования можно сказать что HH работает не только по России, но и в других страннах. Лидеры вакансий являются крупные города как Москва, Санк-Петербург. Самыми востребованными специалистами работающие в режиме "Полная занятость" по графику "Полный день и Удаленная работа".

2. Зарплатная вилка составляет от 71 тыс. - 110 тыс. рублей. Работодателями, предлагающими большую часть вакансий являются крупнейшие российские корпорации: "Яндекс", "Ростелеком", "Тинькофф", "СБЕР", "Газпром нефть".

3. К профессии Data Science приходится 3.6% вакансий от общего числа. Требования довольно жесткие, должен обладать более 5 профессиональными навыками для работы с данными.

4. В общем и целом можно сказать, что данный анализ является поверхностным. Но согласно международным отчетам по будующим профессиям на ближайщие 10 лет, самыми востребованными являются: Data Science, Специалисты по анализу данных, ML, AI. Таким образом, данная область востребована на многие годы.

In [49]:
connection.close()